include/picongpu/particles/atomicPhysics/stateRepresentation/ConfigNumber.hpp

```diff
@@ -0,0 +1,333 @@
+                 *
+                 * @tparam T_DataType unsigned integer data type to represent the config number
+                 * @tparam T_numberLevels max principle quantum number, n_max, represented in
+                 *                              this configNumber
+                 * @tparam T_atomicNumber ... atomic number of the ion, in units of elementary charge
+                 *
+                 * public methods:
+                 *   - T_DataType numberStates()
+                 *       return number of states representable in this configNumber Object
+                 *   - ConfigNumber( uint x )
+                 *       constructor directly setting private member configNumber,
+                 *       BEWARE: no conversions and no runtime range checks, only use if you know
+                 *           what you are doing
+                 *   - ConfigNumber( pmacc::math::Vector<uint_8, T_numberLevels > N )
+                 *       constructor using occupation number vector N = ( N_1, N_2, ..., N_(n_max) )
+                 *   - T_DataType getAtomicStateConfigNumber( pmacc::math::Vector<uint8_t, T_numberLevels> N )
+                 *       converts occupation number vector N to configNumber index
+                 *   - pmacc::math::Vector<uint8_t, T_numberLevels> getLevelVector( T_DataType x )
+                 *       converts configNumber ID x to occupation number vector
+                 */
+                template<typename T_DataType, uint8_t T_numberLevels, uint8_t T_atomicNumber>
+                class ConfigNumber
+                {
+                private:
+                    //! actual storage configNumber ID
+                    T_DataType configNumber;
+
+                    /** g(n) = n^2, maximum possible occupation number of the n-th shell
+                     *
+                     * returns the maximum possible occupation number for the n-th shell,
+                     *
+                     * or in other words how many different electron states there are for a given
+                     * principal quantum number n.
+                     *
+                     * @param n principal quantum number
+                     */
+                    HDINLINE static constexpr uint16_t g(uint8_t n)
+                    {
+                        // cast necessary to prevent overflow in n^2 calculation
+                        return (static_cast<uint16_t>(n) * static_cast<uint16_t>(n) * 2);
+                    }
+
+                    /** number of different occupation number values possible for the n-th shell
+                     *
+                     * BEWARE: n larger than 254 cause an overflow.
+                     *
+                     * @param n principal quantum number
+                     */
+                    HDINLINE static uint16_t numberOfOccupationNumberValuesInShell(uint8_t n)
+                    {
+                        PMACC_ASSERT_MSG(n < 255, "n too large, must be < 255");
```

Comment:
must be `PMACC_DEVICE_ASSERT_MSG()` because this assert is called on device

Reply:
In cases where you compile with `pic-build -c "-DCMAKE_BUILD_TYPE=Debug"` this code should most likely not be compileable.

Reply:
done